### PR TITLE
wip: refactor Validation System (plus message provision and async validation) 

### PR DESCRIPTION
**Pull request by tlouisse**
Working proposal for the following topics:

### Override validation messages
  - [Open the demo](http://localhost:9001/?path=/story/forms-input-string-validation--new-override-messages-per-instance)

```html
<lion-input
  .validators="${[new EqualsLength(4, { message: ({ fieldName }) => `4 ${fieldName} chars please..`' })]}"
  .modelValue="${'123'}"
  label="Custom message for validator instance">
</lion-input>
```

### Custom validators should be easier

- [Open the demo](https://deploy-preview-271--lion-web-components.netlify.com/?path=/story/forms-input-string-validation--new-custom-validator)

```js
    class MyValidator extends Validator {
      constructor(...args) {
        super(...args);
        this.name = 'myValidator';
      }

      execute(modelValue, param) {
        return modelValue === param;
      }

      static getMessage({ fieldName, modelValue, validatorParams: param }) {
        return 'my validator msg';
      }
    }
```

### Async validation support
  - [Open the demo](https://deploy-preview-271--lion-web-components.netlify.com/?path=/story/forms-input-string-validation--new-async)
```js
    class AsyncValidator extends Validator {
      constructor(...args) {
        super(...args);
        this.name = 'asyncValidator';
        this.async = true;
      }

      async execute(modelValue) {
        console.log('async pending...');
        await fetch('something');
        console.log('async done...');
        return modelValue !== 'forbidden';
      }
      ...
    }
```
###  Decouple localize system from ValidateMixin
```js
// set messages globally
Required.getMessage = async data => {
  await validateNamespace;
  return localize.msg('lion-validate:error.required', data);
};
```
### Reconsider validation types
  - Create a generic, flexible feedback system that also allows to create the “twitter max length validation”
_Should we set in stone the 4 validators we currently have? Or extend the concept to a more generic concept ‘Conditional feedback’ that accepts multistate instead of Boolean states. Practical example: a textarea max length almost fits the infoValidator concept, but for instance in ing-web it would have wrong styling preconfigured. It might be better to let people create their own feedback types instead of trying to cover many possible scenarios in the core. Errorvalidators will always be the default_
      - [Open the demo](https://deploy-preview-271--lion-web-components.netlify.com/?path=/story/forms-input-string-validation--new-override-messages-per-instance)

### Improve api
  - Required is a bit odd (could also be a validator you import instead of the string)

  - Validator instances should have state:
        - Are my translations loaded? (if not, I need to (instruct to) rerender validation feedback when so. More info: https://github.com/ing-bank/lion/issues/90
        - Do my parameters change? If so, rerun validations
          - [Open the demo](https://deploy-preview-271--lion-web-components.netlify.com/?path=/story/forms-input-string-validation--new-dynamic-parameter-changes)

- **Identify platform ‘gaps’ and provide alternative solutions**
```js
export class Required extends Validator {
  constructor(...args) {
    super(...args);
    this.name = 'required';
  }

  // eslint-disable-next-line class-methods-use-this
  onFormControlConnect(formControl) {
    if (formControl.inputElement) {
      formControl.inputElement.setAttribute('aria-required', 'true');
    }
  }

  // eslint-disable-next-line class-methods-use-this
  onFormControlDisconnect(formControl) {
    if (formControl.inputElement) {
      formControl.inputElement.removeAttribute('aria-required');
    }
  }
};
```

- **Provide good examples in storybook (purely for validation)**
- **be backwards compatible (?)** 
        - Create a compat layer on our extension layer

### CODE REVIEW

**Comment by tlouisse** [comment]
This mr is a replacement for the previous found here: https://github.com/ing-bank/lion/pull/270

File: packages/validate/test/ValidateCoreMixin.test.js [hunk 1]
```diff
@@ -0,0 +1,1178 @@
...
+      const el = await fixture(html`
+        <${tag} .validators=${[new AlwaysValid(), new AlwaysValid(null, { async: true })]}>
+          ${lightDom}
+        </${tag}>
+      `);
+      const syncSpy = sinon.spy(el, '__executeSyncValidators');
+      const asyncSpy = sinon.spy(el, '__executeAsyncValidators');
+      el.modelValue = 'nonEmpty';
+      expect(syncSpy).to.be.calledBefore(asyncSpy);
+    });
```
Comment (daKmoR) [inline packages/validate/test/ValidateCoreMixin.test.js]: should also allow a way to do something like `await el.validationDone;`?

File: packages/validate/test/ValidateCoreMixin.test.js [hunk 2]
```diff
@@ -0,0 +1,1178 @@
...
+
+      const syncSpy = sinon.spy(el2, '__executeSyncValidators');
+      const resultSpy2 = sinon.spy(el2, '__executeResultValidators');
+
+      el2.modelValue = 'nonEmpty';
+      expect(syncSpy).to.be.calledBefore(resultSpy2);
+    });
+
+    describe('Finalization', () => {
+      it('fires "validation-done" event', async () => {});
```
Comment (daKmoR) [inline packages/validate/test/ValidateCoreMixin.test.js]: maybe rename it to `async-validation-done` as "normal" are sync always..

File: packages/validate/test/ValidateCoreMixin.test.js [hunk 3]
```diff
@@ -0,0 +1,1178 @@
...
+      expect(el.hasAttribute('is-pending')).to.be.true;
+
+      asyncVResolve();
+      await aTimeout();
+      expect(el.isPending).to.be.false;
+      expect(el.hasAttribute('is-pending')).to.be.false;
+    });
+
+    // TODO: run these methods without actually waiting debounce time?
+    it('debounces async validation for performance', async () => {
```
Comment (daKmoR) [inline packages/validate/test/ValidateCoreMixin.test.js]: skip for now

File: packages/validate/test/ValidateCoreMixin.test.js [hunk 4]
```diff
@@ -0,0 +1,1178 @@
...
+      `);
+      // debounce started
+      el.validators = [asyncV];
+      expect(asyncVAbortSpy.called).to.equal(0);
+      el.modelValue = 'dogger';
+      // await original debounce period...
+      expect(asyncVAbortSpy.called).to.equal(1);
+    });
+
+    it('developer can configure condition for asynchronous validation', async () => {
```
Comment (daKmoR) [inline packages/validate/test/ValidateCoreMixin.test.js]: skip for now

File: packages/validate/test/ValidateCoreMixin.test.js [hunk 5]
```diff
@@ -0,0 +1,1178 @@
...
+          }
+        },
+      );
+      const defaultRequiredTag = unsafeStatic(defaultRequiredTagString);
+
+      const el = await fixture(html`
+      <${defaultRequiredTag}
+        .validators=${[new MinLength(3)]}
+      >${defaultRequiredTag}</${tag}>`);
+
```
Comment (daKmoR) [inline packages/validate/test/ValidateCoreMixin.test.js]: make `_preconfiguredValidators` public => so people could override messages?
```js
 const el = await fixture(html`
      <${tag}
        .validators=${[new MinLength(3)]}
        .preconfiguredValidators=${[new Required(null, { message: () => return 'My Error message' })]}
      ></${tag}>`);
```
or individually `el.preconfiguredValidators[0].options.message` ???? hmm these are classes? so maybe not?

add 
```js
get allValidators() {
  return [...this.preconfiguredValidators, ...this.validators];
}
```

so `.validators` stays pure to what the users sets

File: packages/validate/test/ValidateMixin.test.js [hunk 6]
```diff
@@ -1030,19 +1055,50 @@ describe.only('ValidateMixin', () => {
...
+      expect(el._feedbackNode.message).to.equal('');
+      el.validators = [new AlwaysInvalid()];
+      await el.feedbackComplete;
+      expect(el._feedbackNode.message).to.equal('Message for alwaysInvalid');
+    });
+
     it('has configurable feedback visibility hook', async () => {
       const el = await fixture(html`
         <${tag}
-          ._prioritizeAndFilterFeedback=${() => []}
```
Comment (daKmoR) [inline packages/validate/test/ValidateMixin.test.js]: so we do not want to allow overriding it within the template? e.g. only when extending? then we should also not test it via
```
el._prioritizeAndFilterFeedback = () => [];
```
?

File: packages/validate/src/loadDefaultFeedbackMessages.js [hunk 7]
```diff
@@ -5,6 +5,18 @@ import { DefaultSuccess } from './resultValidators/DefaultSuccess.js';
 
 let loaded = false;
 
+/**
+ * @desc bridges between ValidateMixin (that only supplies formControl) and translations
+ * (that know about less technical concepts)
+ */
+function transformData(data) {
```
Comment (daKmoR) [inline packages/validate/src/loadDefaultFeedbackMessages.js]: nice 👍

**Comment by daKmoR** [comment]
this has been addressed via a different pull request 👍 

if something is still open feel free to reopen or create another PR or issue 🤗

**Pull request closed without merging**